### PR TITLE
[AIRFLOW-5758] Support the custom cursor classes for the PostgreSQL hook

### PR DESCRIPTION
**Pull request by mingrammer**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-5758
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

**Comment by mingrammer** [comment]
There is a failed test case.

```
16) FAIL: test_wait_for_response_running (tests.gcp.hooks.test_kubernetes_engine.TestGKEClusterHook)
----------------------------------------------------------------------
   Traceback (most recent call last):
    /usr/local/lib/python3.6/unittest/mock.py line 1183 in patched
      return func(*args, **keywargs)
    tests/gcp/hooks/test_kubernetes_engine.py line 309 in test_wait_for_response_running
      self.assertEqual(time_mock.call_count, 3)
   AssertionError: 11202 != 3
   -------------------- >> begin captured logging << --------------------
   airflow.gcp.hooks.kubernetes_engine.GKEClusterHook: INFO: Waiting for OPERATION_NAME <Mock name='mock.name' id='140681303781840'>
   --------------------- >> end captured logging << ---------------------
```

File: airflow/hooks/postgres_hook.py [hunk 1]
```diff
@@ -67,6 +68,13 @@ def get_conn(self):
             password=conn.password,
             dbname=self.schema or conn.schema,
             port=conn.port)
+        if conn.extra_dejson.get('cursor', False):
+            if (conn.extra_dejson['cursor']).lower() == 'dictcursor':
+                conn_args['cursor_factory'] = psycopg2.extras.DictCursor
+            elif (conn.extra_dejson['cursor']).lower() == 'realdictcursor':
+                conn_args['cursor_factory'] = psycopg2.extras.RealDictCursor
+            elif (conn.extra_dejson['cursor']).lower() == 'namedtuplecursor':
+                conn_args['cursor_factory'] = psycopg2.extras.NamedTupleCursor
```
Comment (feluelle) [inline airflow/hooks/postgres_hook.py]: This looks like a [factory pattern](https://en.wikipedia.org/wiki/Factory_method_pattern) can be of good use here :)

```python
def _get_cursor(self):
    _cursor = conn.extra_dejson['cursor'].lower()
    if _cursor == 'dictcursor':
        return psycopg2.extras.DictCursor
    if _cursor == 'realdictcursor':
        return psycopg2.extras.RealDictCursor
    if _cursor == 'namedtuplecursor':
        return psycopg2.extras.NamedTupleCursor
    raise ValueError('Invalid cursor passed {}'.format(_cursor))
```

I also added complete validation (see the last line).

So after you have this factory you can just call it like that.
```python
if conn.extra_dejson.get('cursor', False):
    conn_args['cursor_factory'] = self._get_cursor()
```

WDYT?

Comment (mingrammer) [reply airflow/hooks/postgres_hook.py]: That is a great pointing. I'd love to apply your suggestions.

And, I think we can also use this pattern for the mysql hook as well.

Comment (mingrammer) [reply airflow/hooks/postgres_hook.py]: Are there some conventions for the order of public/private methods definitions in a class?

Comment (mingrammer) [reply airflow/hooks/postgres_hook.py]: Hm, wait. `conn` is a local variable for `get_conn` method. So we should get the connection in the `_get_cursor` again by

```python
conn_id = getattr(self, self.conn_name_attr)
conn = self.get_connection(conn_id)
```

So `get_connection` will be called twice when cursor class is specified. Is it ok?

Or we can pass the `conn.extra_dejson['cursor']` to the `_get_cursor` from `get_conn`.

Comment (feluelle) [reply airflow/hooks/postgres_hook.py]:
```python
def _get_cursor(self, raw_cursor):
    _cursor = raw_cursor.lower()
    if _cursor == 'dictcursor':
        return psycopg2.extras.DictCursor
    if _cursor == 'realdictcursor':
        return psycopg2.extras.RealDictCursor
    if _cursor == 'namedtuplecursor':
        return psycopg2.extras.NamedTupleCursor
    raise ValueError('Invalid cursor passed {}'.format(_cursor))
```

```python
raw_cursor = conn.extra_dejson.get('cursor', False)
if raw_cursor:
    conn_args['cursor_factory'] = self._get_cursor(raw_cursor)
```

Comment (mingrammer) [reply airflow/hooks/postgres_hook.py]: I updated the code to use the cursor factory and added an additional test case.

**Review by feluelle — APPROVED** [review]

**Comment by feluelle** [comment]
Hey @mingrammer do you also want to add a note to the class docstring about the functionality you added (like the notes about ssl and AWS IAM auth.) ? This way users will have it easier to use it.

**Comment by mingrammer** [comment]
Sure, I'll do!

**Comment by mingrammer** [comment]
Done.

**Comment by mingrammer** [comment]
Some azure-dependant test cases are failing due to azure library changes.

**Comment by feluelle** [comment]
I see. Let's wait until https://github.com/apache/airflow/pull/6486 is merged and then I rerun the failed ones.

**Comment by feluelle** [comment]
Thanks @mingrammer :)

**Pull request merged**